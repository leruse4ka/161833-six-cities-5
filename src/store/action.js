export const ActionType = {
  CHANGE_CITY: `CHANGE_CITY`,
  CHANGE_SORT: `CHANGE_SORT`,
  CHANGE_ACTIVE_ID: `CHANGE_ACTIVE_ID`,
  RESET_ACTIVE_ID: `RESET_ACTIVE_ID`,
};

export const ActionCreator = {
  changeCity: (evt) => {
    evt.preventDefault();
    return {
      type: ActionType.CHANGE_CITY,
      payload: evt.target.textContent,
    };
  },
  changeSort: (evt) => {
    evt.preventDefault();
    return {
      type: ActionType.CHANGE_SORT,
      payload: evt.target.textContent,
    };
  },
  focusActiveId: (offer) => {
    return {
      type: ActionType.CHANGE_ACTIVE_ID,
      payload: offer.id,
    };
  },
  resetActiveId: () => {
    return {
      type: ActionType.RESET_ACTIVE_ID,
      payload: null,
    };
  }
};

