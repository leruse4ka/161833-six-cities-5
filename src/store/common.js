export const getErr = (err) => {
  const form = document.querySelector(`form`);
  form.innerHTML += `<div class="errComments" style="width: 300px; height: 100px; color: red; font-size: 25px; position: absolute; top: 80%; left: 50%; background-color: white">${err}</div>`;


  const handleClickErr = (evt) => {
    evt.preventDefault();
    document.querySelector(`.errComments`).remove();
    window.removeEventListener(`click`, handleClickErr);
  };

  window.addEventListener(`click`, handleClickErr);
};
