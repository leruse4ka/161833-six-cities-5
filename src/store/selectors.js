import {createSelector} from "reselect";
import {sorting, sortingComments} from "../sorting";

const getOffers = (state) => state.DATA.offers;
const getCurrentSort = (state) => state.STATUS.currentSort;
const getCityName = (state) => state.STATUS.cityName;

export const getFilterOffers = createSelector(
    getOffers,
    getCurrentSort,
    getCityName,
    (offers, currentSort, cityName) => sorting(currentSort, Array.from(offers.slice().filter((item) => item.city.name === cityName)))
);

const getComments = (state) => state.DATA.comments;

export const getSortComments = createSelector(
    getComments,
    (comments) => sortingComments(comments)
);
