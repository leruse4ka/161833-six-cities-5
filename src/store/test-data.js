export const mockOffer = {
  bedrooms: 3,
  city: {
    location: {
      latitude: 52.370216,
      longitude: 4.895168,
      zoom: 10
    },
    name: `Amsterdam`
  },
  description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
  inside: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
  host: {
    avatar: `img/1.png`,
    id: 3,
    isPro: true,
    name: `Angelina`
  },
  id: 1,
  photoGallery: [`img/1.png`, `img/2.png`],
  isFavorite: false,
  premium: false,
  location: {
    latitude: 52.35514938496378,
    longitude: 4.673877537499948,
    zoom: 8
  },
  adults: 4,
  previewImage: `img/1.png`,
  price: 120,
  rating: 4.8,
  title: `Beautiful & luxurious studio at great location`,
  type: `apartment`
};

export const noop = () => {};

export const getMockOffers = (count) => {
  const mockOffers = Array(count)
    .fill(``)
    .map((_, index) => {
      return {
        bedrooms: 3,
        city: {
          location: {
            latitude: 52.370216,
            longitude: 4.895168,
            zoom: 10
          },
          name: `Amsterdam`
        },
        description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
        inside: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
        host: {
          avatar: `img/1.png`,
          id: 3,
          isPro: true,
          name: `Angelina`
        },
        id: index + 1,
        photoGallery: [`img/1.png`, `img/2.png`],
        isFavorite: false,
        premium: false,
        location: {
          latitude: 52.35514938496378,
          longitude: 4.673877537499948,
          zoom: 8
        },
        adults: 4,
        previewImage: `img/1.png`,
        price: 120,
        rating: 4.8,
        title: `Beautiful & luxurious studio at great location`,
        type: `apartment`
      };
    });

  return mockOffers;
};

export const commentsMock = [
  {
    comment: `Comment`,
    date: `2020-10-30T19:03:49.647Z`,
    id: 1,
    rating: 3,
    user: {
      avatar: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/4.jpg`,
      id: 13,
      isPro: false,
      name: `La`,
    }
  },
  {
    comment: `Comment`,
    date: `2020-10-30T19:03:49.647Z`,
    id: 2,
    rating: 5,
    user: {
      avatar: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/7.jpg`,
      id: 12,
      isPro: false,
      name: `ABC`,
    }
  },
];

export const loginDataMock = {
  "avatar_url": `img/1.png`,
  "email": `Oliver.conner@gmail.com`,
  "id": 1,
  "is_pro": false,
  "name": `Oliver.conner`
};

export const adaptedLoginDataMock = {
  avatar: `img/1.png`,
  email: `Oliver.conner@gmail.com`,
  id: 1,
  isPro: false,
  name: `Oliver.conner`
};
