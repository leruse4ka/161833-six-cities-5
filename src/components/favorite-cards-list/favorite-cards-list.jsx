import React from "react";
import PropTypes from "prop-types";
import mainPageOffersProp from "../pages/main-page/main-page-offers.prop";
import Card from "../card/card";
import {DefaultType} from "../../const";
import {connect} from "react-redux";
import {ActionCreator} from "../../store/action";

const FavoriteCardsList = (props) => {
  const {offers, cityName, onOfferFocus, onOfferLeave} = props;
  return (
    offers
      .filter((offer) => offer.city.name === cityName)
      .map((el) => <Card
        offer={el}
        key={el.id}
        defaultType={DefaultType.favorites}
        onOfferFocus={onOfferFocus}
        onOfferLeave={onOfferLeave}
      />)
  );
};

FavoriteCardsList.propTypes = {
  offers: mainPageOffersProp,
  cityName: PropTypes.string.isRequired,
  onOfferFocus: PropTypes.func.isRequired,
  onOfferLeave: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  onOfferFocus(offer) {
    dispatch(ActionCreator.focusActiveId(offer));
  },
  onOfferLeave() {
    dispatch(ActionCreator.resetActiveId());
  }
});

export default connect(null, mapDispatchToProps)(FavoriteCardsList);
