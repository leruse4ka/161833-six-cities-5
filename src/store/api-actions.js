import {ActionCreator} from "./action";
import {AuthorizationStatus} from "../const";
import {adaptToClient, adaptToServer, adaptToClientComments} from "../utils";

export const fetchOffersList = () => (dispatch, _getState, api) => (
  api.get(`/hotels`)
    .then(({data}) => dispatch(ActionCreator.loadOffers(data.map(adaptToClient))))
);

export const checkAuth = () => (dispatch, _getState, api) => (
  api.get(`/login`)
    .then(() => dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH)))
    .catch(() => {})
);

export const fetchOffer = (id) => (dispatch, _getState, api) => (
  api.get(`/hotels/${id}`)
    .then(({data}) => dispatch(ActionCreator.loadOffer(adaptToClient(data))))
    .then(() => dispatch(ActionCreator.redirectToRoute(`/offer/${id}`)))
);

export const fetchOffersFavorite = () => (dispatch, _getState, api) => (
  api.get(`/favorite`)
    .then(({data}) => dispatch(ActionCreator.loadOffersFavorite(data.map(adaptToClient))))
);

export const fetchOffersNearbyList = (id) => (dispatch, _getState, api) => (
  api.get(`/hotels/${id}/nearby`)
    .then(({data}) => dispatch(ActionCreator.loadOffersNearby(data.map(adaptToClient))))
);

export const fetchComments = (id) => (dispatch, _getState, api) => (
  api.get(`/comments/${id}`)
    .then(({data}) => dispatch(ActionCreator.loadOffersComments(data.map(adaptToClientComments))))
);

export const login = ({login: email, password}) => (dispatch, _getState, api) => (
  api.post(`/login`, {email, password})
    .then(() => dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH)))
    .then(() => dispatch(ActionCreator.redirectToRoute(`/favorites`)))
);

export const setFavorite = (id, status, offer) => (_dispatch, _getState, api) => (
  api.post(`favorite/${id}/${status}`, adaptToServer(offer))
);

export const comments = (id, {comment, rating}) => (dispatch, _getState, api) => (
  api.post(`comments/${id}`, {comment, rating})
    .then(() => dispatch(ActionCreator.submitComment()))
);
