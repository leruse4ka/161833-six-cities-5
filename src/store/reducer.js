import {extend} from "../utils";
import {ActionType} from "./action";
import offers from "../mocks/offers";

const initialState = {
  cityName: `Paris`,
  offers,
  favoriteOffers: offers.filter((item) => item.isFavorite),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY:
      return extend(state, {
        cityName: action.payload,
      });
  }

  return state;
};

export {reducer};
