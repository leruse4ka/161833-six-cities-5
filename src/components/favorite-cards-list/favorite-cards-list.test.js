import React from "react";
import renderer from "react-test-renderer";
import {FavoriteCardsList} from "./favorite-cards-list";
import {noop, getMockOffers} from "../../store/test-data";
import {BrowserRouter} from "react-router-dom";

it(`Render FavoriteCardsList`, () => {
  const tree = renderer.create(
      <BrowserRouter>
        <FavoriteCardsList
          offers={getMockOffers(3)}
          cityName={`Amsterdam`}
          onOfferFocus={noop}
          onOfferLeave={noop}
          onFavoriteClick={noop}
          onOfferClick={noop}
        >
          <React.Fragment />
        </FavoriteCardsList>
      </BrowserRouter>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
