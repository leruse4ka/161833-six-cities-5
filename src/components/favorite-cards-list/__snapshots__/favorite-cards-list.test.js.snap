// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render FavoriteCardsList 1`] = `
Array [
  <article
    className="favorites__card place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    
    <div
      className="favorites__image-wrapper place-card__image-wrapper"
    >
      <a
        href="/offer/1"
        onClick={[Function]}
      >
        <img
          alt="Place image"
          className="place-card__image"
          height={110}
          src="img/1.png"
          width={150}
        />
      </a>
    </div>
    <div
      className="favorites__card-info place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            120
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              href="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            To
             bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "100%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        <a
          href="/offer/1"
          onClick={[Function]}
        >
          Beautiful & luxurious studio at great location
        </a>
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>,
  <article
    className="favorites__card place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    
    <div
      className="favorites__image-wrapper place-card__image-wrapper"
    >
      <a
        href="/offer/2"
        onClick={[Function]}
      >
        <img
          alt="Place image"
          className="place-card__image"
          height={110}
          src="img/1.png"
          width={150}
        />
      </a>
    </div>
    <div
      className="favorites__card-info place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            120
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              href="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            To
             bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "100%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        <a
          href="/offer/2"
          onClick={[Function]}
        >
          Beautiful & luxurious studio at great location
        </a>
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>,
  <article
    className="favorites__card place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    
    <div
      className="favorites__image-wrapper place-card__image-wrapper"
    >
      <a
        href="/offer/3"
        onClick={[Function]}
      >
        <img
          alt="Place image"
          className="place-card__image"
          height={110}
          src="img/1.png"
          width={150}
        />
      </a>
    </div>
    <div
      className="favorites__card-info place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            120
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              href="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            To
             bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "100%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        <a
          href="/offer/3"
          onClick={[Function]}
        >
          Beautiful & luxurious studio at great location
        </a>
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>,
]
`;
